### PR TITLE
Adjust GOJA to benchmark data

### PR DESCRIPTION
**Pull request by flukson**

### CODE REVIEW

File: goja/EventAnalysis.cpp [hunk 1]
```diff
@@ -52,6 +87,50 @@ void EventAnalysis::select_coincident_hits(vector<Hit> &hits) {
 
 }
 
+void EventAnalysis::select_coincident_singles(vector<Hit> &hits) {
+
+  bool debug = true; // TODO: rmove
```
Comment (kamilrakoczy) [inline goja/EventAnalysis.cpp:92]: Should this be removed before PR?

**Review by kamilrakoczy — APPROVED** [review]
Generally LGTM, just 1 comment.

**Review by kamilrakoczy — APPROVED** [review]

File: goja/EventAnalysis.cpp (hunk 1 above)
Comment (flukson) [reply goja/EventAnalysis.cpp:92]: Yes. The code was not finished. In the new version this line is removed.

**Review by kamilrakoczy — APPROVED** [review]

File: goja/EventAnalysis.cpp [hunk 2]
```diff
@@ -29,6 +33,38 @@ void sort_hits(vector<Hit> &hits, string key) {
 
 }
 
+Hit add_hits(Hit &h1, Hit &h2, string winner="centroidWinner") {
```
Comment (wkrzemien) [inline goja/EventAnalysis.cpp:36]: I don't understand why you consider only two hits?
If you have more than 2 hits, then centroidWinner is just wrong in this implemenetation

Comment (wkrzemien) [reply goja/EventAnalysis.cpp:36]: For energy weighting see e.g.:
https://github.com/JPETTomography/J-PET-geant4/blob/master/Objects/Geant4/DetectorHit.cpp

Comment (wkrzemien) [inline goja/EventAnalysis.cpp:36]: I guess it would be better to define `enum AveragingMethod` and define `kCentroidWinner` etc

File: goja/Hit.h [hunk 3]
```diff
@@ -11,10 +11,9 @@ class Hit {
 public :
 
   int eventID;          // ID of the event
-  int volumeID;         // ID of the detecting volume
-  int trackID;          // ID of the track
-  int parentID;         // ID of the parent particle
-  int primaryID;        // ID of the primary particle
+  int volumeID;         // volume detection in 'scanner' GATE systemType
```
Comment (wkrzemien) [inline goja/Hit.h:14]: You can initialize the members here e.g. `int volumeID = -1;`

**Comment by flukson** [comment]
@wkrzemien I have fixed the energy weigthing and added GATE-like approach.

File: goja/EventAnalysis.h [hunk 4]
```diff
@@ -22,7 +22,17 @@ enum EventType {
...
+enum AveragingMethod {
+  kUnspecifiedAM = 0,
+  kCentroidWinnerNaivelyWeighted = 1,
+  kCentroidWinnerEnergyWeighted = 2,
+  kCentroidWinnerEnergyWeightedFirstTime = 3,
+  kEnergyWinner = 4
+};
+
 void sort_hits(vector<Hit> &hits, string key);
+Hit add_hits(const Hit &h1, const Hit &h2, const AveragingMethod winner);
```
Comment (wkrzemien) [inline goja/EventAnalysis.h:34]: Is this method needed?

File: goja/EventAnalysis.h [hunk 5]
```diff
@@ -35,9 +45,10 @@ public :
   EventAnalysis();
 
   void select_coincident_hits(vector<Hit> &hits);
+  void select_coincident_singles(const std::vector<Hit> &hits);
   EventType verify_type_of_coincidence(Hit &h1, Hit &h2);
   void print_coincidences();
-  void analyze_event(vector<Hit> &hits);
+  void analyze_event(vector<Hit> &hits, bool singles = true);
```
Comment (wkrzemien) [inline goja/EventAnalysis.h:51]: It is not necessary, but I would change the name of the boolean to e.g. `areSingles` or something similar.

File: goja/Hits.C [hunk 6]
```diff
@@ -14,7 +14,7 @@ using namespace std;
 #include "EventAnalysis.h"
 #include "Common.h"
 
-LoopResults Hits::Loop() {
+LoopResults Hits::Loop(bool singles) {
```
Comment (wkrzemien) [inline goja/Hits.C:17]: As above `areSingles` or `hits_are_singles` sounds better.

**Review by wkrzemien — COMMENTED** [review]
The algos look fine for me. Good job. Two small comments.

File: goja/EventAnalysis.cpp [hunk 7]
```diff
@@ -29,6 +34,84 @@ void sort_hits(vector<Hit> &hits, string key) {
 
 }
 
+Hit add_hits(const std::vector<Hit> &hits, const AveragingMethod winner = kCentroidWinnerEnergyWeightedFirstTime) {
+
+  /*
```
Comment (wkrzemien) [inline goja/EventAnalysis.cpp:39]: I guess this you can already remove

**Review by wkrzemien — COMMENTED** [review]
Nice work! I have only one comment  + one request -> Please, add some kind of description about the available options e.g. reading singles, reading hits etc.

File: goja/EventAnalysis.cpp [hunk 8]
```diff
@@ -29,6 +34,84 @@ void sort_hits(vector<Hit> &hits, string key) {
...
+  const unsigned int multiplicity = hits.size();
+  double x1 = hits[0].posX;
+  double y1 = hits[0].posY;
+  double z1 = hits[0].posZ;
+  double x2 = hits[multiplicity-1].posX;
+  double y2 = hits[multiplicity-1].posY;
+  double z2 = hits[multiplicity-1].posZ;
+  double distance = sqrt((x1-x2)*(x1-x2) + (y1-y2)*(y1-y2) + (z1-z2)*(z1-z2));
+  std::cout << multiplicity << '\t' << distance << std::endl;
+  */
```
Comment (wkrzemien) [inline goja/EventAnalysis.cpp:49]:
```suggestion

```

Comment (wkrzemien) [inline goja/EventAnalysis.cpp]:
```suggestion

```

**Review by wkrzemien — APPROVED** [review]
To speedup I leave this one small remark unsolved and I merge it

**Pull request merged**